<div class="alert alert-info">
Привет, Ольга! Меня зовут Светлана Чих и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>

**SQL**

**Цель исследования** - проанализировать базу данных сервиса для чтения книг по подписке и сформулировать ценностное предложение для нового продукта.

**Объект исследования** - база данных сервиса для чтения книг по подписке. 

1. [Исследование таблиц](#1)
2. [Задача 1. Посчитайте, сколько книг вышло после 1 января 2000 года](#2)
3. [Задача 2. Для каждой книги посчитайте количество обзоров и среднюю оценку](#3)
4. [Задача 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры](#4)
5. [Задача 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками](#5)
6. [Задача 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#6)
7. [Выводы](#7)

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<a id="1"></a> **1. Исследование таблиц** 

In [2]:
query = ''' SELECT *
            FROM books
            LIMIT 3
        ''' 
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [3]:
query = ''' SELECT *
            FROM authors
            LIMIT 3
        ''' 
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

In [4]:
query = ''' SELECT *
            FROM publishers
            LIMIT 3
        ''' 
pd.io.sql.read_sql(query, con = engine)

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

In [5]:
query = ''' SELECT *
            FROM ratings
            LIMIT 3
        ''' 
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [6]:
query = ''' SELECT *
            FROM reviews
            LIMIT 3
        ''' 
pd.io.sql.read_sql(query, con = engine)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

<a id="2"></a> **2. Посчитайте, сколько книг вышло после 1 января 2000 года**

In [7]:
query = ''' SELECT COUNT(title)
            FROM books
            WHERE publication_date > '01.01.2000'
        ''' 
pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

<a id="3"></a> **3. Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [8]:
query = ''' SELECT b.book_id,
                   b.title, 
                   ROUND(AVG(rating),1) AS rating_avg,
                   COUNT(DISTINCT text) AS review_count
            FROM books AS b
            LEFT JOIN reviews AS r ON b.book_id = r.book_id
            LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
            GROUP BY b.book_id
            -- HAVING ROUND(AVG(rating),1)  >=4
            -- ORDER BY rating_avg DESC, review_count DESC
            -- LIMIT 10
        ''' 
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  rating_avg  \
0          1                                       'Salem's Lot         3.7   
1          2                 1 000 Places to See Before You Die         2.5   
2          3  13 Little Blue Envelopes (Little Blue Envelope...         4.7   
3          4  1491: New Revelations of the Americas Before C...         4.5   
4          5                                               1776         4.0   
..       ...                                                ...         ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)         3.7   
996      997                        Xenocide (Ender's Saga  #3)         3.4   
997      998                                    Year of Wonders         3.2   
998      999                        You Suck (A Love Story  #2)         4.5   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...         3.8   

     review_count  
0               2  
1               1  
2               3  
3               2  
4               4  
..            ...  
995             3  
996             3  
997             4  
998             2  
999             4  

[1000 rows x 4 columns]

Для каждой книги посчитали количество обзоров и среднюю оценку. Результат отсортировали по убыванию среднего рейтинга.

 <div class="alert alert-danger">
 <s><b>😔 Необходимо исправить:</b> А почему не все книги вошли? В задаче нет никаких условий по рейтингу</s>
 </div>

 <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Исправила. Подумала, что нужно результат задачи под цель исследования подвести и сделала чуть больше :) </div>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

<a id="4"></a> **4. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [9]:
query = ''' SELECT t.publisher
            FROM
            (
                 SELECT p.publisher_id,
                        p.publisher,
                        count(title)  
                 FROM publishers AS p
                 LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
                 WHERE b.num_pages > 50
                 GROUP BY p.publisher_id
                 ORDER BY count DESC
             ) AS t    
            LIMIT 1
        ''' 
pd.io.sql.read_sql(query, con = engine)

publisher
0  Penguin Books

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

<a id="5"></a> **5. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [10]:
query = '''SELECT tt.author
           FROM 
               (
                SELECT t.author,     
                AVG(rating) AS rating_avg
                FROM 
                    (
                     SELECT  a.author,
                             b.title,
                             COUNT(rt.rating) OVER (PARTITION BY a.author, b.title) AS rating_count,
                             rt.rating
                     FROM author AS a
                     LEFT JOIN books AS b ON a.author_id = b.author_id
                     LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
                  
                     ) AS t
               
                WHERE rating_count >= 50
                GROUP BY t.author 
                ORDER BY rating_avg DESC
                ) AS tt
           
           LIMIT 1
        ''' 
pd.io.sql.read_sql(query, con = engine)

author
0  J.K. Rowling/Mary GrandPré

Среди книг с 50 и более оценками и с самой высокой средней оценкой книг лидирует автор J.K. Rowling/Mary GrandPré.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

<a id="6"></a> **6. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок**

In [11]:
query = '''SELECT ROUND(AVG(tt.review_count)) AS review_avg
           FROM 
              (
               SELECT username,
                      COUNT(text) AS review_count
               FROM reviews
               WHERE username IN
                  (
                   SELECT t.username
                   FROM
                      (
                       SELECT username,
                              COUNT(rating) AS rating_count
                       FROM ratings 
                       GROUP BY username
                       ) AS t
                   WHERE t.rating_count > 50 
              
                  ) 
           GROUP BY username   
              ) AS tt
                               
                         '''
pd.io.sql.read_sql(query, con = engine)

review_avg
0        24.0

Пользователи, которые поставили больше 50 оценок, делают в среднем 24 обзора.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

<a id="7"></a> **7. Вывод**

- выяснили, что после 1 января 2000 года вышло всего 819 книг.

- выявили список книг со средними рейтингами и количеством отзывов, что позволит фильтровать книги с наилучшими показателями: максимальным средним рейтингом и максимальным количествои отзывов.

- определили издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books - издательство, с которым выгодно сотрудничать.

- среди наиболее популярных книг (с 50 и более оценками и с самой высокой средней оценкой) выявили лидера - это автор J.K. Rowling/Mary GrandPré. Кники данного автора наиболее продаваемые.

- самые активные пользователи (которые поставили больше 50 оценок) делают в среднем 24 обзора, что позволяет другим читателям выьирать и покупать книги.


<div class="alert alert-success">
 <b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! У тебя получились уверенные работы на финальном спринте, логичные и последовательные, ты умешшь рассуждать, визуализировать данные и делать выводы, это очень хорошие качества для аналитика! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/profile/ycloud-datalens/'> DataLens<a>
 </div>